docs/root/operations/tools/route_table_check_tool.rst

```diff
@@ -18,9 +18,15 @@ Usage
       Schema: All internal schemas in the tool are based on :repo:`proto3 <test/tools/router_check/validation.proto>`.
 
     -c <string>,  --config-path <string>
-      Path to a router config file (YAML or JSON). The router config file schema is found in
-      :ref:`config <envoy_v3_api_file_envoy/config/route/v3/route.proto>` and the config file extension
-      must reflect its file type (for instance, .json for JSON and .yaml for YAML).
+       Path to a router config file (YAML or JSON). The router config file schema is found in
+       :ref:`config <envoy_v3_api_file_envoy/config/route/v3/route.proto>` and the config file extension
+       must reflect its file type (for instance, .json for JSON and .yaml for YAML).
+
+    --bootstrap-config <string>
```

Comment:
The main issue I see here is that it is unintuitive to have either a route or bootstrap+listener.
I think it would be better if there's either a validation of all the routes in the bootstrap, or a given route.
Requiring a listener seems to be a non-required constraint. WDYT?

Reply:
> The main issue I see here is that it is unintuitive to have either a route or bootstrap+listener.

I understand the concern, I kept routes flag avoid a breaking change to existing users of the tool.

> I think it would be better if there's either a validation of all the routes in the bootstrap, or a given route.
Requiring a listener seems to be a non-required constraint. WDYT?

I think we'd still need to be more specific when reading the whole configuration instead of just the route config as the same input (e.g authority) might match different routes in different listeners and the test case might match an undesired route. For example:

```yaml
tests:
- test_name: Test acme external cluster
  input:
    authority: www.acme.org
    path: /
    method: GET
  validate:
    cluster_name: outbound|80|external|acme.org
```

```yaml
  listeners:
    - name: internal
      address:
        socket_address:
          address: 0.0.0.0
          port_value: 10000
      filter_chains:
        - filters:
            - name: envoy.filters.network.http_connection_manager
              typed_config:
                "@type": "type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager"
                stat_prefix: internal
                route_config:
                  name: internal
                  virtual_hosts:
                    - name: default
                      domains:
                        - acme.org
                      routes:
                        - match:
                            prefix: "/"
                          route:
                            cluster: outbound|80|internal|acme.org
                http_filters:
                  - name: envoy.filters.http.router
                    typed_config:
                      "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
    - name: external
      address:
        socket_address:
          address: 0.0.0.0
          port_value: 10001
      filter_chains:
        - filters:
            - name: envoy.filters.network.http_connection_manager
              typed_config:
                "@type": "type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager"
                generate_request_id: true
                tracing: {}
                stat_prefix: external
                route_config:
                  name: external
                  virtual_hosts:
                    - name: default
                      domains:
                        - acme.org
                      routes:
                        - match:
                            prefix: "/"
                          route:
                            cluster: outbound|80|external|acme.org
                http_filters:
                  - name: envoy.filters.http.router
                    typed_config:
                      "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
```

Another option would be to read from cluster_dump configuration but we'd still need to specify the route name to find the route in `type.googleapis.com/envoy.admin.v3.RoutesConfigDump`

Reply:
I suppose the route name (or listener name) to use could also be part of the test input:

```yaml
tests:
- test_name: Test acme external cluster
  input:
    authority: www.acme.org
    path: /
    method: GET
    route_name: external
  validate:
    cluster_name: outbound|80|external|acme.org
```

Reply:
> I understand the concern, I kept routes flag avoid a breaking change to existing users of the tool.

Keeping it is good, otherwise it will break current users.

> I think we'd still need to be more specific when reading the whole configuration instead of just the route config as the same input (e.g authority) might match different routes in different listeners and the test case might assert against an undesired route

Wouldn't one expect to have all the routes in a given bootstrap to be validated?

> I suppose the route name (or listener name) to use could also be part of the test input:

Why is this needed?

Reply:
> Wouldn't one expect to have all the routes in a given bootstrap to be validated?

Yes. I'm sorry for the ambiguity on this thread. The tool as is implemented today takes a single file through `-c` in a [config.route.v3.RouteConfiguration](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route.proto) format and the tests runs against this single RouteConfiguration. When reading the whole Envoy configuration, either from [Bootstrap](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/bootstrap/v3/bootstrap.proto) as currently implemented in this PR through `--bootstrap-config` or [ConfigDump](https://www.envoyproxy.io/docs/envoy/latest/api-v3/admin/v3/config_dump.proto) as another option, we could have more than one [config.route.v3.RouteConfiguration](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route.proto) present in the file (for example in different listeners) and should be tested separately ([see](https://github.com/envoyproxy/envoy/pull/35662#discussion_r1761726314) use case above). When running the test we need to instruct the tool which [config.route.v3.RouteConfiguration](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route.proto) to use to run the test against.

We can't use/combine all [config.route.v3.RouteConfiguration](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route.proto) as each RouteConfiguration should have its own set of tests.

> Why is this needed?

To find which RouteConfiguration to run the test against. I'm happy to jump in a call or something to clarify.

Reply:
Thanks for clarifying. I'm just a bit worried that this option will confuse more than it will help.
Specifically I'm wondering what should be done with listener configs that have filters that update/modify the headers. This may impact the routing decision, and so it is not clear to me that validating on the listener level is intuitive in this case.

If I were to do it, I would probably create a different tool that does the bootstrap validation, as this seems to me to be quite different than the a single route check.
That said, this may be too much work for a use-case that might not be too useful.

> We can't use/combine all [config.route.v3.RouteConfiguration](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route.proto) as each RouteConfiguration should have its own set of tests.

You are saying that this needs to be a property of the test, not of the execution.
If so, doesn't it make more sense to have the listener name in the validation-input of the test?

Reply:
> You are saying that this needs to be a property of the test, not of the execution.
If so, doesn't it make more sense to have the listener name in the validation-input of the test?

That's what I meant in https://github.com/envoyproxy/envoy/pull/35662#discussion_r1761754300

> Specifically I'm wondering what should be done with listener configs that have filters that update/modify the headers

I understand the concerns. My initial goal loading the whole configuration was a step towards https://github.com/envoyproxy/envoy/issues/12393 and https://github.com/envoyproxy/envoy/issues/13065 while fixing my runtime flags issue (https://github.com/envoyproxy/envoy/issues/35174), but it seems a bigger change (if ever done). Should we reduce the scope to load runtime flags instead? Do you have a suggestion how to configure the tool with certain flags? e.g cli arguments, validation input.

Reply:
@adisuissa friendly ping.

Reply:
Thank you @adisuissa. I think on the Validation level would make lots of sense. I will close this PR and follow-up on a new one limited to runtime flags.